runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SdkComponents.java

```diff
@@ -191,6 +193,10 @@ private String getApplicationName(AppliedPTransform<?, ?, ?> appliedPTransform)
     if (name.isEmpty()) {
       name = "unnamed-ptransform";
     }
+    // Normalize, trim, and uniqify.
+    int maxNameLength = 100;
+    name = Normalizer.normalize(name, Form.NFC).replaceAll("[^A-Za-z0-9-_]", "-");
```

Comment:
Let's document the required regex in the pipeline proto. Is there validation logic like there is in Python? That would be a good place to check it, too.

Reply:
Added a comment to the proto recommending alphanumeric characters, '_', and '-' for transform IDs.
Also added a test.

Corresponding Python changes are in https://github.com/apache/beam/pull/14502.